File: native/cocos/scene/Ambient.cpp
/****************************************************************************
 Copyright (c) 2021-2023 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
****************************************************************************/

#include "scene/Ambient.h"
#include "core/Root.h"
#include "renderer/pipeline/PipelineSceneData.h"
#include "renderer/pipeline/RenderPipeline.h"
#include "renderer/pipeline/custom/RenderInterfaceTypes.h"

namespace {
cc::Color col;

// Normalize HDR color
void normalizeHDRColor(cc::Vec4 &color) {
    const float intensity = 1.F / std::max(std::max(std::max(color.x, color.y), color.z), 0.0001F);
    if (intensity < 1.F) {
        color.x *= intensity;
        color.y *= intensity;
        color.z *= intensity;
    }
}

} // namespace
namespace cc {
namespace scene {

void AmbientInfo::setSkyLightingColor(const Color &val) {
    Vec4 v4(static_cast<float>(val.r) / 255.F, static_cast<float>(val.g) / 255.F, static_cast<float>(val.b) / 255.F, static_cast<float>(val.a) / 255.F);
    const bool isHDR = Root::getInstance()->getPipeline()->getPipelineSceneData()->isHDR();
    if (isHDR) {
        _skyColorHDR.set(v4);
    } else {
        _skyColorLDR.set(v4);
    }
    if (_resource != nullptr) {
        _resource->setSkyColor(v4);
    }
}

const Color &AmbientInfo::getSkyLightingColor() const {
    const bool isHDR = Root::getInstance()->getPipeline()->getPipelineSceneData()->isHDR();
    Vec4 v4(isHDR ? _skyColorHDR : _skyColorLDR);
    normalizeHDRColor(v4);
    col.set(static_cast<uint8_t>(v4.x * 255.F), static_cast<uint8_t>(v4.y * 255.F), static_cast<uint8_t>(v4.z * 255.F), 255);
    return col;
}

void AmbientInfo::setSkyColor(const Vec4 &val) {
    const bool isHDR = Root::getInstance()->getPipeline()->getPipelineSceneData()->isHDR();
    if (isHDR) {
        _skyColorHDR.set(val);
    } else {
        _skyColorLDR.set(val);
    }
    if (_resource != nullptr) {
        _resource->setSkyColor(val);
    }
}

void AmbientInfo::setSkyIllum(float val) {
    const bool isHDR = Root::getInstance()->getPipeline()->getPipelineSceneData()->isHDR();
    if (isHDR) {
        _skyIllumHDR = val;

    } else {
        _skyIllumLDR = val;
    }

    if (_resource != nullptr) {
        _resource->setSkyIllum(val);
    }
}

float AmbientInfo::getSkyIllum() const {
    const bool isHDR = Root::getInstance()->getPipeline()->getPipelineSceneData()->isHDR();
    return isHDR ? _skyIllumHDR : _skyIllumLDR;
}

void AmbientInfo::setGroundLightingColor(const Color &val) {
    const Vec4 v4(static_cast<float>(val.r) / 255.F, static_cast<float>(val.g) / 255.F, static_cast<float>(val.b) / 255.F, static_cast<float>(val.a) / 255.F);
    const bool isHDR = Root::getInstance()->getPipeline()->getPipelineSceneData()->isHDR();
    if (isHDR) {
        _groundAlbedoHDR.set(v4);
    } else {
        _groundAlbedoLDR.set(v4);
    }

    if (_resource != nullptr) {
        _resource->setGroundAlbedo(v4);
    }
}

const Color &AmbientInfo::getGroundLightingColor() const {
    const bool isHDR = Root::getInstance()->getPipeline()->getPipelineSceneData()->isHDR();
    Vec4 v4(isHDR ? _groundAlbedoHDR : _groundAlbedoLDR);
    normalizeHDRColor(v4);
    col.set(static_cast<uint8_t>(v4.x * 255.F), static_cast<uint8_t>(v4.y * 255.F), static_cast<uint8_t>(v4.z * 255.F), 255);
    return col;
}

void AmbientInfo::setGroundAlbedo(const Vec4 &val) {
    const bool isHDR = Root::getInstance()->getPipeline()->getPipelineSceneData()->isHDR();
    if (isHDR) {
        _groundAlbedoHDR.set(val);
    } else {
        _groundAlbedoLDR.set(val);
    }

    if (_resource != nullptr) {
        _resource->setGroundAlbedo(val);
    }
}

void AmbientInfo::activate(Ambient *resource) {
    _resource = resource;
    _resource->initialize(this);
}

//
void Ambient::initialize(AmbientInfo *info) {
    // setSkyColor(info->getSkyColor());
    // setGroundAlbedo(info->getGroundAlbedo());
    // _skyIllum = info->getSkyIllum();

    // Init HDR/LDR from serialized data on load
    _skyColorHDR = info->getSkyColorHDR();
    _groundAlbedoHDR.set(info->getGroundAlbedoHDR());
    _skyIllumHDR = info->getSkyIllumHDR();
    _skyColorLDR = info->getSkyColorLDR();
    _groundAlbedoLDR.set(info->getGroundAlbedoLDR());
    _skyIllumLDR = info->getSkyIllumLDR();
}

Vec4 &Ambient::getSkyColor() {
    const bool isHDR = Root::getInstance()->getPipeline()->getPipelineSceneData()->isHDR();
    return isHDR ? _skyColorHDR : _skyColorLDR;
}

const Vec4 &Ambient::getSkyColor() const {
    const bool isHDR = Root::getInstance()->getPipeline()->getPipelineSceneData()->isHDR();
    return isHDR ? _skyColorHDR : _skyColorLDR;
}

void Ambient::setSkyColor(const Vec4 &color) {
    const bool isHDR = Root::getInstance()->getPipeline()->getPipelineSceneData()->isHDR();
    if (isHDR) {
        _skyColorHDR.set(color);
    } else {
        _skyColorLDR.set(color);
    }
}

/**
 * @en Sky illuminance
 * @zh 天空亮度
 */
float Ambient::getSkyIllum() const {
    const bool isHDR = Root::getInstance()->getPipeline()->getPipelineSceneData()->isHDR();
    return isHDR ? _skyIllumHDR : _skyIllumLDR;
}
void Ambient::setSkyIllum(float illum) {
    const bool isHDR = Root::getInstance()->getPipeline()->getPipelineSceneData()->isHDR();
    if (isHDR) {
        _skyIllumHDR = illum;
    } else {
        _skyIllumLDR = illum;
    }
}

/**
 * @en Ground color
 * @zh 地面颜色
 */
const Vec4 &Ambient::getGroundAlbedo() const {
    const bool isHDR = Root::getInstance()->getPipeline()->getPipelineSceneData()->isHDR();
    return isHDR ? _groundAlbedoHDR : _groundAlbedoLDR;
}

void Ambient::setGroundAlbedo(const Vec4 &color) {
    const bool isHDR = Root::getInstance()->getPipeline()->getPipelineSceneData()->isHDR();
    if (isHDR) {
        _groundAlbedoHDR.set(color);
    } else {
        _groundAlbedoLDR.set(color);
    }
}

} // namespace scene
} // namespace cc

File: cocos/render-scene/scene/ambient.ts
/*
 Copyright (c) 2020-2023 Xiamen Yaji Software Co., Ltd.

 https://www.cocos.com/

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
*/

import { Vec4, cclegacy } from '../../core';
import { AmbientInfo } from '../../scene-graph/scene-globals';

/**
 * @en Ambient lighting representation in the render scene.
 * The initial data is setup in [[SceneGlobals.ambient]].
 * @zh 渲染场景中的环境光照设置。
 * 初始值是由 [[SceneGlobals.ambient]] 设置的。
 */
export class Ambient {
    /**
     * @en Default sun illuminance
     * @zh 默认太阳亮度
     */
    public static SUN_ILLUM = 65000.0;
    /**
     * @en Default sky illuminance
     * @zh 默认天空亮度
     */
    public static SKY_ILLUM = 20000.0;

    /**
     * @en Enable ambient
     * @zh 是否开启环境光
     */
    set enabled (val: boolean) {
        this._enabled = val;
    }
    get enabled (): boolean {
        return this._enabled;
    }
    /**
     * @en Sky color
     * @zh 天空颜色
     */
    get skyColor (): Vec4 {
        const isHDR = (cclegacy.director.root).pipeline.pipelineSceneData.isHDR;
        if (isHDR) {
            return this._skyColorHDR;
        } else {
            return this._skyColorLDR;
        }
    }
    set skyColor (color: Vec4) {
        const isHDR = (cclegacy.director.root).pipeline.pipelineSceneData.isHDR;
        if (isHDR) {
            this._skyColorHDR.set(color);
        } else {
            this._skyColorLDR.set(color);
        }
    }

    /**
     * @en Sky illuminance
     * @zh 天空亮度
     */
    get skyIllum (): number {
        const isHDR = (cclegacy.director.root).pipeline.pipelineSceneData.isHDR;
        if (isHDR) {
            return this._skyIllumHDR;
        } else {
            return this._skyIllumLDR;
        }
    }
    set skyIllum (illum: number) {
        const isHDR = (cclegacy.director.root).pipeline.pipelineSceneData.isHDR;
        if (isHDR) {
            this._skyIllumHDR = illum;
        } else {
            this._skyIllumLDR = illum;
        }
    }
    /**
     * @en Ground color
     * @zh 地面颜色
     */
    get groundAlbedo (): Vec4 {
        const isHDR = (cclegacy.director.root).pipeline.pipelineSceneData.isHDR;
        if (isHDR) {
            return this._groundAlbedoHDR;
        } else {
            return this._groundAlbedoLDR;
        }
    }
    set groundAlbedo (color: Vec4) {
        const isHDR = (cclegacy.director.root).pipeline.pipelineSceneData.isHDR;
        if (isHDR) {
            this._groundAlbedoHDR.set(color);
        } else {
            this._groundAlbedoLDR.set(color);
        }
    }

    protected _groundAlbedoHDR = new Vec4(0.2, 0.2, 0.2, 1.0);
    protected _skyColorHDR = new Vec4(0.2, 0.5, 0.8, 1.0);
    protected _skyIllumHDR = 0;

    protected _groundAlbedoLDR = new Vec4(0.2, 0.2, 0.2, 1.0);
    protected _skyColorLDR = new Vec4(0.2, 0.5, 0.8, 1.0);
    protected _skyIllumLDR = 0;

    protected _mipmapCount = 1;

    protected _enabled = false;

    public initialize (ambientInfo: AmbientInfo): void {
        // Init HDR/LDR from serialized data on load
        this._skyColorHDR = ambientInfo.skyColorHDR;
        this._groundAlbedoHDR.set(ambientInfo.groundAlbedoHDR);
        this._skyIllumHDR = ambientInfo.skyIllumHDR;

        this._skyColorLDR = ambientInfo.skyColorLDR;
        this._groundAlbedoLDR.set(ambientInfo.groundAlbedoLDR);
        this._skyIllumLDR = ambientInfo.skyIllumLDR;
    }
}

cclegacy.Ambient = Ambient;
